### ISSUE TITLE
Add "reverse predict" method

### ISSUE THREAD

**Issue by spudzee** [fchollet/keras#857]
There should be a simple mechanism for reversing the order of Keras neural networks. So if you have input X and output Y, you could completely reverse the network and have input Y and output X; keeping the optimized weights.

It might be helpful for hyperparameter optimization to see the neural network's idea of various targets. For instance, in classifying MNIST digits you could get an idea of how the neural network perceives various digits.

**Comment by bayesrule** [comment]
This is quite a valuable function. This "feedback" / "inverse" mechanism is widely discussed and studied in Learning theories.

**Comment by fchollet** [comment]
As I'm sure you are aware, it is very much a non-trivial problem as many NN operations are non-reversible. It definitely isn't implementable in the general case and we won't seek to implement it in Keras. If you want to set it up for simple networks (e.g. fully connected), it's easy to do with Keras on your own (without even modifying Keras, simply via weight transfer).

> you could get an idea of how the neural network perceives various digits.

The way this is usually done is by gradient ascent in the input space.

**Comment by spudzee** [comment]
Alright fair enough :) I didn't realize the details were so complicated.
